pkg/reconciler/openshift/tektonconfig/rbac.go

```diff
@@ -393,78 +394,102 @@ func (r *rbac) createResources(ctx context.Context) error {
-		if err != nil {
-			withError = true
-			logger.Errorf("failed to bind scc to namespace %s, %v", ns.Name, err)
-		}
+			var withError bool
 
-		if sa != nil {
-			// We use a namespace scoped Role when SCC annotation is present, and
-			// a cluster scoped ClusterRole when the default SCC is used
-			roleRef := r.getSCCRoleInNamespace(&ns)
-			if roleRef != nil {
-				if err := r.ensurePipelinesSCCRoleBinding(ctx, sa, roleRef); err != nil {
-					withError = true
-					logger.Errorf("failed to create Pipeline Scc Role Binding in namespace %s, %v", ns.Name, err)
-				}
-			}
-			if err := r.ensureRoleBindings(ctx, sa); err != nil {
+			logger.Infow("Inject CA bundle configmap in ", "Namespace", ns.GetName())
+			if err := r.ensureCABundles(ctx, &ns); err != nil {
 				withError = true
-				logger.Errorf("failed to create rolebinding in namespace %s, %v", ns.Name, err)
+				logger.Errorf("failed to ensure ca bundles presence in namespace %s, %v", ns.Name, err)
 			}
 
-			if err := r.ensureClusterRoleBindings(ctx, sa); err != nil {
+			// If "operator.tekton.dev/scc" exists in the namespace, then bind
+			// that SCC to the SA
+			if err := r.handleSCCInNamespace(ctx, &ns); err != nil {
 				withError = true
-				logger.Errorf("failed to create clusterrolebinding in namespace %s, %v", ns.Name, err)
+				logger.Errorf("failed to bind scc to namespace %s, %v", ns.Name, err)
 			}
-		}
-		if !withError {
-			logger.Infof("add label namespace-reconcile-version to mark namespace '%s' as reconciled", ns.Name)
 
-			// Re-fetch the namespace to ensure we have the latest version
-			updatedNS, err := r.kubeClientSet.CoreV1().Namespaces().Get(ctx, ns.Name, metav1.GetOptions{})
+			logger.Infow("Ensures Default SA in ", "Namespace", ns.GetName())
+			sa, err := r.ensureSA(ctx, &ns)
 			if err != nil {
-				return fmt.Errorf("failed to re-fetch namespace %s: %v", ns.Name, err)
+				withError = true
+				logger.Errorf("failed to ensure default SA in namespace %s, %v", ns.Name, err)
 			}
 
-			// Get the current labels, or initialize if none exist
-			nsLabels := updatedNS.GetLabels()
-			if nsLabels == nil {
-				nsLabels = map[string]string{}
+			if sa != nil {
+				// We use a namespace scoped Role when SCC annotation is present, and
+				// a cluster scoped ClusterRole when the default SCC is used
+				roleRef := r.getSCCRoleInNamespace(&ns)
+				if roleRef != nil {
+					if err := r.ensurePipelinesSCCRoleBinding(ctx, sa, roleRef); err != nil {
+						withError = true
+						logger.Errorf("failed to create Pipeline Scc Role Binding in namespace %s, %v", ns.Name, err)
+					}
+				}
+				if err := r.ensureRoleBindings(ctx, sa); err != nil {
+					withError = true
+					logger.Errorf("failed to create rolebinding in namespace %s, %v", ns.Name, err)
+				}
+
+				// Locking section to ensure clusterRoleBinding is updated by one goroutine at a time
+				mu.Lock()
+				if err := r.ensureClusterRoleBindings(ctx, sa); err != nil {
```

Comment:
Won't this generate *a lot* of API calls (possibly, on get and one update or create) where we could gather the namespaces and update the `ClusterRoleBinding` at the end (no matter if some namespaces failed — we would only use the ones that didn't faiil)

Reply:
Revised the code to address the comment by collecting all subjects and making a single call to update the ClusterRoleBinding, rather than performing a read/update for each subject on per namespace basis.
[TOBE.pdf](https://github.com/user-attachments/files/18212602/TOBE.pdf)
[AS_IS_Flow.pdf](https://github.com/user-attachments/files/18212604/AS_IS_Flow.pdf)